# Bài 00. Machine Learning là gì ?

**Machine learning (máy học)** được dùng để lấy tri thức từ dữ liệu. Nó là một lĩnh vực nghiên cứu của các lĩnh vực: Statistics(Thống kê), Artificial Intelligence (trí tuệ nhân tạo) và Computer Science(Khoa học máy tính). Các ứng dụng của Machine Learning (Applications) trở nên phổ biến và không còn xa vời như trước, càng tiến dần về tương lai thì ứng dụng của máy học trong cuộc sống hằng ngày phổ rộng và hầu như ở đâu cũng có máy móc thay thế vai trò lao động của con người.

Điểm qua một số ứng dụng trong cuộc sống: hệ thống khuyến nghị tự động về thức ăn, quần áo, nhận dạng con người trong một bức hình, trong các trang web Google, Facebook, Amazone, Netflix thì khối lượng AI, Machine learning rất lớn(hầu như tính năng nào cũng có tích hợp AI),
Ngòai các ứng dụng trong thương mại, Machine Learling là công cụ trong các nghiên cứu các dữ liệu lớn ngày nay.

## Nội dung:
1. **Why Machine Learning ?**
2. **Why Python ?**
3. **Phân loại hoa lan**
4. **Kết luận**
----

## 1. Why Machine Learning ?

### Cuộc sống là một chuỗi các bài toán cần ta giải quyết ! :)

![Complex Prolems](https://sheilapontis.files.wordpress.com/2013/03/106-problemsolving1.jpg)

Với nhu cầu của con người ngày nay, từ "ăn chắc mặc bền" rồi "ăn no mặt ấm" , "ăn ngon mặt đẹp" rồi muốn gì được đó, thì tri thức của con người phải đủ lớn, đủ khả năng đáp ứng các bài toán không phải đơn giản như trước. 

Ngày xưa, người ta tìm cách giải phương trình để tìm nghiệm đúng nhất. Phải qua hàng thế kỉ mơí có các cách giải. Đến ngày nay, ta tìm nghiệm của phương trình bậc 2 hay bậc 3, nếu là con người thì chắc vài phút, nhanh thì vài giây; nếu là máy tính thì bằng một cái click chuột hoặc là một tổ hợp phím nào đó. Thật sự các bài toán trong thế giới thực ngày nay không thể giải bằng một bộ não đơn thuần của con người. Nhưng nếu hợp nhất nhiều bộ não thì sao, làm sao được ? Phải tìm ra một vật chất gì đó thay thế cho việc học có giới hạn của con người, và phải làm sao cho nó gặp phương trình nào cũng giải được.

Câu trả lời đó là là máy tính. Vì máy tính có tốc độ xử lí nhanh, không bị ảnh hưởng bởi các điều kiện ngoại vi, khả năng lưu trữ rất lớn về dữ liệu. Nhưng vấn đề được đặt ra là làm sao làm cho cái máy học được ? Làm sao làm cho nó có tư duy, cho nó biết phân biệt đúng sai... Thế nên, mấy chục năm trước đây, đã có lĩnh vực mang tên gọi là Artificial Intelligence (Trí tuệ nhân tạo), **this is the research field aiming to how to make machine have the same intelligence human**. Giúp cho cái máy tính, chiếc xe ô tô có thể nhìn như con người (Computer Vision), nghe và viết như con người (Natural Language Processing),...rồi từ việc học nhanh, học nhiều, xử lý nhanh vốn có, ta dùng nó giải quyết bài toán phức tạp, rắc rối mà ta không có khả năng giải quyết bằng sức người. Vấn đề lại được đặt ra, làm sao cho cái máy điều "phi thường" (do ta không bao giờ nghĩ rằng nó đạt được như vậy) như vậy, liệu chuyện "cái máy biết học" có điên rồ quá không ?    

![Characteristics of Computer](https://www.freezonal.com/files/2010/10/Characteristics-of-Computer.gif)

Để làm được những điều thần thánh như trên, ta phải có một công cụ đủ mạnh để thực hiện điều đó. Không phải là sử dụng máy tính cực mạnh, tốc độ xử lý nhanh là được mà phải làm sao dạy học cho máy tính. Machine Learning chính là công cụ, phương thức để cho máy tính có thể học tập, qua công cụ này ta sẽ dạy cho máy tính cách tiếp nhận tri thức, xử lí vấn đề theo cách nào đó gần giống với con người, nhưng với sức mạnh của máy tính, không bị tác động ảnh hưởng bởi điều kiện ngoại cảnh, cảm xúc, lí trí.

Xét về tính liên quan, **Machine Learning là một tập con của trí tuệ nhân tạo (Artificial Intelligence), nhưng cũng đóng vai trò là phương tiện để Trí tuệ nhân tạo đạt được mục tiêu đề ra**. Còn một cách định nghĩa chính xác về mặc ngữ nghĩa thì được biểu diễn như sau:

![Presentation of Machine Learning](https://i0.wp.com/www.zldoty.com/wp-content/uploads/2017/04/2017-04-06-001-Machine-Learning-Definition-ETP-Framework.png?resize=1003%2C550&fbclid=IwAR0m0TH17okIlcEf3U1i2wfCoRm0JBcr7k2Pv8WGEkVziQmgTlrr2P2KNQY)

Đương nhiên Machine  Learnign chỉ là một công cụ, mà đã là công cụ hay một sự vật hiện tượng thì phải có hạn chế. Đừng nghĩ về hạn chế, hãy nghĩ về những gì mà nó có thể làm cho con người.

### *Những bài toán mà Machine Learning có thể giải quyết*

Những bài toán có thể giải quyết hay nói đúng hơn là các bài toán nào có thể cho máy tính học tập rồi sau này nó sẽ giải quyết bằng chính sức của nó. Có rất nhiều bài toán mà một máy tính hay một hệ thống có thể giải quyết. Và việc học này về cơ bản có hai hình thức là **học có dạy bảo** và học **không dạy bảo** hay thế giới ngày nay gọi là **supervised learning** and **unsupervised learning**. 

#### Supervised Learning(Học có giám sát)
Nói một cách nôm na là đưa cho một đứa trẻ một tấm hình chứa con mèo, cô giáo nói với nó là có con mèo trong đó, tiếp tục lại đưa cho nó một tấm hình, cô giáo lại bảo là có con mèo trong tấm hình, tiếp tục lần thứ ba nhưng cô giáo khẳng định là không có mèo trong hình,... Quá trình cô giáo dạy em bé cách nhận biết trong hình có con mèo hay không chính là việc learning từ những ví dụ mẫu, dần dần em bé có khả năng nhận biết được trong một bức ảnh lạ lẫm bất kỳ có con mèo hay không(chưa chắc lúc nào bé nhận định cũng đúng nhưng thường thì đúng). Thực sự là như vậy, nhưng theo định nghĩa thì **từ tập training data (dữ liệu huấn luyện) được gán nhãn hãy xây dựng một mô hình có thể dự đoán output mà ta quan tâm sao cho output này thật là giống với điều hiển nhiên**.

#### Unsupervised Learning(Học không giám sát)

- **Nếu như học mà không ai giám sát(ở đây không phải là quản lí) thì ta phải tự bơi !** Khi đưa cho đứa trẻ một tập hợp rất là nhiều hình nhưng chỉ có đúng là 3 hình dạng là tròn, vuông và tam giác(đừng nói cho đứa trẻ tên gọi của các hình, chỉ nói cho bé là có 3 loại hình). Hỏi rằng bé có tự chia một đống hình đó thành 3 loại hình dạng hay không ? Tương tự như vậy, **từ tập dữ liệu huấn luyện đầu vào (training data input), khác với supervised learning, học không giám sát không có tập nhãn cho các dữ liệu, hãy xây dựng một mô hình để dự đoán, đưa ra thông tin mà ta quan tâm**.

Ngoài ra còn các thuật tóan trong Machine  Learning khác:
- Học tăng cường
- Học bán giám sát
- ...

![Machine Learning Algorithms](https://media.geeksforgeeks.org/wp-content/cdn-uploads/20190522174744/MachineLearning.png)

Nghe sơ qua thấy chũng dễ dễ và cũng khó khó. Dễ vì chưa hiểu rõ cái cách cũng như quy trình và khó ở chỗ chưa bao giờ tiếp cận với những điều mình chưa bao giờ nghĩ. Hãy điểm qua một số điều hấp dẫn của Machine Learning đem lại trong thời gian gần đây nào:

- Traffic Alerts (Maps)
- Social Media (Facebook)
- Transportation and Commuting (Uber)
- Products Recommendations
- Virtual Personal Assistants
- Self Driving Cars
- Dynamic Pricing
- Google Translate
- Online Video Streaming (Netflix)

*Có thể tham khảo chi tiết hơn tại [đây](https://www.edureka.co/blog/machine-learning-applications/).*

![Applications of Machine Learning](https://i0.wp.com/www.hopetutors.com/wp-content/uploads/2019/06/Machine-Learning-Training.png)

**Để áp dụng được Machine Learning vào các bài toán thì phải có nền tảng kiến thức nhân loại, cụ thể ở đây là Toán học và Lập trình. Đem toán vào trong máy tính, lập trình cho máy tính thực hiện thuật toán xây dựng mô hình tử dữ liệu đầu vào.**

Nói về toán học, để biết cách xử dụng  Machine Learning, và hiểu được cách xây dựng Machien Learning model thì phạm vi toán học bao gồm có Đại số tuyến tính(Linear Algebra), Giải tích Ma trận (Matrix Calculus), Xác xuất thống kê(Probability and Statistics Mathematical), Tối ưu (Approximations),... 

## 2. Why Python ?

### Quá nhiều ngôn ngữ lập trình !

![Histogram of Programming Languages](https://pbs.twimg.com/media/EBbmxOGU4AA_CqL.jpg)

Nếu toán học rộng đến như vậy, thì có vẽ lập trình cũng khá là rất rối đấy.  **Lập trình để đưa các mô hình toán học xuất phát từ bài toán thực tế vào máy tính đó là một vấn đề (Cuộc sống là một chuỗi các problems)**. Và phải chọn ngôn ngữ để lập trình. Bạn có thể chọn C/C++, Java, Ruby, Python, ...Nhưng do Python là một ngôn ngữ truyền thống cho việc lập trình về Toán, Machine Learning, xử lý các bài toán phức tạp. Xin nói thêm một chút, Python thì Open Source Code, có cộng đồng lập trình viên sử dụng đông đảo, các ứng đụng viết bằng ngôn ngữ Python rất phổ biến, và hầu như bây giờ là thời kì của Python and its applications.

Điểm qua các thư viện hỗ dùng nhiều trong Machine Learning cũng như các lĩnh vực mở rộng:
- **Numpy**: Numerical Python - Đây là thư viện hỗ trợ vector nhiều chiều, tính toán số học, biến đổi Furier, ...
- **Matplotib**: thư viện phục vụ cho việc visualize dữ liệu.
- **Scikit-learn**: nơi tập hợp các thuật toán trong Machine Learning.
- **Pandas**: hỗ trợ cho việc phân tích dữ liệu, tính toán khoa học.

![Libabries for Data Science](https://lasseschultebraucks.com/assets/pythonStack.jpg)

### Lập trình trên môi trường nào khi làm Machine Leanring 

Chọn được một môi trường để lập trình cho một ngôn ngữ cụ thể có thể rất mất sức. Làm sao chọn môi trường phù hợp với máy mình, lại phù hợp với điều mình làm, balabala,.... Ở đây, theo như thế giới ngày nay thì hầu như người ta sử dụng Jupyter Notebook, Google Colab để học tập cũng như làm việc với Machine Learning, vì các môi trường này lưu lại kết quả sau khi thực thi, cũng như có thể note lại những tâm tư, nguyện vọng, tình cảm, trí tuệ trong lúc học tập. Ngoài ra, còn có các IDE khác, lập trình Python cũng êm: Pycharm, IDLE,...

### Jupyter Notebook

![Jupyter Notebook Icon](http://5b0988e595225.cdn.sohucs.com/images/20190815/828ec608c7f64784af516f433b4c0e4b.jpeg)

**`Jupyter notebook is a very popular and flexible tool which lets us put our code, output of the code and any kind of visualization or plot etc. in the same document.`**

Đây là một môi trường dành cho việc lập trình, bên cạnh đó hỗ trợ cho việc ghi chép kết quả và các nội dung diễn giải liên quan đến chương trình, project của bạn. Với môi trường này bạn có thể code ngôn ngữ Python, R, thậm chí là C/C++. Nó rất là thú vị và đáng hứng khởi khi bạn làm việc, bạn có thể ghi lại cảm xúc, suy nghĩ của mình trong lúc code và output của đoạn code mà bạn lập trình sẽ được lưu lại ngay sau bạn run chương trình, dễ dàng cho việc trình bày và xem lại.

Mình đã có làm sẵn một file hướng dẫn cài đặt Jupyter trên Windows và Ubuntu. Hãy dựa theo nó và cài đặt nhé !

## 3. Phân loại hoa lan

Oke chúng ta đã hiểu Machine Learning là gì ? Theo quan điểm của nhiều người mà tôi tán thành thì nó là Toán học mà ta dạy cho máy tính. 

### *Trăm nghe không bằng một thấy*

Nói thì lòng vòng, miên man, hãy cùng xem thử nó như thế nào nhỉ. Ở đây chúng ta cùng lướt qua một ứng dụng kinh điển của Machine  Learning, giúp chúng ta xem được cái hình dạng của Machine Learning, và thấy được Machine Learning là cái gì thực sự.

**Đó là bài toán phân loại hoa Iris (hoa lan)**. Giả đinh rằng bạn là người có kiến thức chuyên môn về hoa. Và tôi đưa cho bạn một bức ảnh vè hoa Iris, yêu cầu bạn cho tôi biết hoa này thuộc dạng hoa Iris nào :) ? Đường nhiên bạn trả lời được ngay. Vậy giờ bạn nghĩ xem, nếu là bạn có chuyên môn thì oke bạn phân dạng hoa được, nếu bạn không có chuyên môn thì tôi sẽ đem một bộ dữ liệu toàn là ảnh chụp về Iris mà tôi có ở nhà để chỉ cho bạn từng tấm coi hoa trong đấy thuộc dạng nào. Dần dần bạn sẽ có được một cái gọi là chuyên môn(kinh nghiệm được học tập từ các ví dụ tôi chỉ). Khi đó quăng cho bạn một tấm hình bất kì, bạn sẽ ngay lập tức phản hồi cho tôi hoa này thuộc dạng nào nào :) .

Oke, Bạn được tôi training, hay nói cách khác bạn học từ những ví dụ mà tôi đã gán nhãn sẵn. Dần dần, bạn trở nên có kinh nghiệm, có chuyên môn và có thể cho ra một output ứng với bức ảnh bất kì về hoa Iris. Tư tưởng như vậy, nhưng hãy nghĩ đến việc đưa nó vào máy tính. Bạn hãy đóng vai trò là tôi, một supervisor, cái máy tính là học viên của bạn, bạn cho nó càng nhiều ví dụ mẫu và chỉ ra nhãn của từng ví dụ. Cuối cùng, cái máy sẽ biết phân loại hoa Iris cho bạn, bạn không cần tốn công nữa, vì bạn đã đem tri thức của bạn dạy lại cho nó và giờ nó sẽ thay thế vai trò của bạn.

![Iris Flower](https://miro.medium.com/max/810/0*Uw37vrrKzeEWahdB)

**Trên đây là 3 dạng hoa Iris. Nếu bạn muốn biết rõ về hoa Iris chũng như có bao nhiêu loại thực sự hãy tham khảo trên Wikipedia, ở đây ta cho là chỉ có 3 dạng hoa này thôi và mục tiêu của ta là làm sao huấn luyện máy tính nhìn vào một bức ảnh sẽ cho ta biết ngay hoa Iris này thuộc loại nào.**

*Có thật sự là mình dạy cho máy tính hiểu được ?*

### *Trăm thấy không bằng một sờ*

Nghe thì tư tưởng và cách tiến hành cũng không khó lắm. Có vẽ sẽ làm được qua lời nói. Nhưng để hiện thực hóa thì phải lập trình vào máy tính. **Đầu tiên là phải có dữ liệu để dạy, sau đó là phải có cách dạy, và cuối cùng là kiêm tra đánh giá coi mình dạy oke chưa**. **Vậy dữ liệu ở đây là gì ? Nó là ảnh, chỉ là ảnh. Hãy đem các ảnh chụp về hoa Iris được chụp để training cho mô hình của chúng ta, nhưng nhớ là mỗi một ảnh phải chỉ rõ là thuộc loại nào nhé. Cách dạy cho nó chính là chọn một thuật toán để training, giống như việc dạy học, phải có một cách gì đó hiệu quả thì học viên tiếp thu mới nhanh và hiệu quả vậy. Việc đánh giá sau khi training chính là đưa cho máy tính một vài bức ảnh, thử xem nó có phân loại đúng hay không ? Dựa vào việc kiểm tra sau khi training ta có thể suy nghĩ lại về cách dạy của ta, cách ta chuẩn bị bài giảng có thực sự tốt hay chưa.** 

*Và sau đây là phần code cho toán bộ lời văn chúng ta đã đọc từ đầu đến giờ. Chỉ vỏn vẹn mười mấy vòng :)*

#### Chuẩn bị công cụ để làm việc

In [0]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn import neighbors, datasets

#### Bước 01. Chuẩn bị dữ liệu

In [0]:
iris = datasets.load_iris()
iris_X = iris.data
iris_y = iris.target
print('Number of classes: %d' %len(np.unique(iris_y)))
print('Number of data points: %d' %len(iris_y))

Number of classes: 3
Number of data points: 150


Đoạn code này có nhiều ý nghĩa nhưng nó đang thực sự làm một việc đó là chuẩn bị bài giảng :). 
- `iris_X` chính là tập ảnh bạn sắp đem đi training cho máy tính
- `iris_y` chính là tập đáp án có sẵn cho mỗi tấm ảnh, hay gọi theo chuyên môn là label của input đầu vào.
- Dựa bào output ta thấy có thực sự là 3 loại hoa(3 class) và số lượng ảnh mà ta có là 150 ảnh (datapoints).

*Xem thử vài mẫu dự liệu nào, xem sơ qua giáo án của chúng ta nhỉ.*

In [0]:
X0 = iris_X[iris_y == 0,:]
print('\nSamples from class 0:\n', X0[:5,:])

X1 = iris_X[iris_y == 1,:]
print('\nSamples from class 1:\n', X1[:5,:])

X2 = iris_X[iris_y == 2,:]
print('\nSamples from class 2:\n', X2[:5,:])


Samples from class 0:
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]

Samples from class 1:
 [[7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.9 1.5]
 [5.5 2.3 4.  1.3]
 [6.5 2.8 4.6 1.5]]

Samples from class 2:
 [[6.3 3.3 6.  2.5]
 [5.8 2.7 5.1 1.9]
 [7.1 3.  5.9 2.1]
 [6.3 2.9 5.6 1.8]
 [6.5 3.  5.8 2.2]]


*Chỗ này không cần quan tâm nhiều lắm, cơ bản mà !*

#### Bước 02. Tách dữ liệu thành hai phần là training data và test data.

Ta phải tách dữ liệu ra ở đây vì phải chừa một phần lại để sau khi dạy cho máy rồi, test lại xem cách dạy của mình như nào và dữ liệu mình training cho nó có phù hợp hay không.

In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(iris_X, iris_y, test_size=50)

print("Training size: %d" %len(y_train))
print("Test size    : %d" %len(y_test))

Training size: 100
Test size    : 50


*Wao.. Như vậy là ta chi tập dữ liệu thành 2 phần: phần training data có 100 ảnh và test data có 50 ảnh. Tức là 100 ảnh sẽ được dùng để làm tài liệu học tập còn 50 ảnh còn lại là thi cuối kì :).*

#### Bước 03. Bắt đầu training nào

In [0]:
clf = neighbors.KNeighborsClassifier(n_neighbors = 1, p = 2)
clf.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

Sau khi đoạn code này thực hiện xong, ta có trong tay một model và model này sẽ được test để đánh giá xem chất lượng như nào.

#### Bước 04. Testing

Quá trình testing như việc bạn chon thử một vài tấm hình, một vài bài tập để kiểm tra xem máy tính mà bạn training từ nảy đến giờ có thực sự học tốt không ?

In [0]:
y_pred = clf.predict(X_test)
print("Print results for 20 test data points:")
print("Predicted labels: ", y_pred[20:40])
print("Ground truth    : ", y_test[20:40])

Print results for 20 test data points:
Predicted labels:  [0 0 2 2 0 0 2 2 2 1 0 1 1 0 2 0 2 0 2 1]
Ground truth    :  [0 0 2 2 0 0 2 2 2 1 0 1 1 0 2 0 2 0 1 2]


Bạn không cần hiểu chi tiết sâu bên trong, hãy quan sát 2 hàng giá trị song song nhau. Gần như các phần tử nằm cùng một vị trí gióng nhau. Điều này có nghĩa là sai sót, hay còn gọi là lỗi của trainer rất ít.

*Đương nhiên nhìn như vậy vẫn chưa mang tính khách quan cho lắm. Phải có số liệu chứ*

In [0]:
from sklearn.metrics import accuracy_score
print("Accuracy of 2NN: %.2f %%" %(100*accuracy_score(y_test, y_pred)))

Accuracy of 2NN: 96.00 %


Waoo... Chính xác đến 96%. Thực sự bạn dạy học rất tốt đáy nhỉ.

**Tuy là bạn không thấy được nó diễn ra như thế nào bằng mắt thường. Nhưng thực sự bạn đã tạo ra được một mô hình có thể phân loại ảnh hoa Iris bất kì thuôc vào 1 trong 3 lớp hoa. Tuy có vẽ không có ý nghĩa gì cho lắ nhưng biết đâu dựa vào đây bạn có thể tạo ra các mô hình có thể phân loại của riêng bạn chẳng hạn vậy :)**

#### Cùng xem lại quá trình chúng ta đi giải quyết một bài toán trong machine learning nào:

![Machine Learning Process](https://onclick360.com/wp-content/uploads/2018/08/ml-process.jpg)

## 4. Kết luận

**Từ đầu tới giờ nói gì nhỉ ??**
- Chúng ta bắt đầu với một giới thiệu ngắn gọn về học máy và các ứng dụng của nó, sau đó thảo luận về sự khác biệt giữa học tập có giám sát và không giám sát và đưa ra một cái nhìn tổng quan về các công cụ mà ta làm việc trong Machine learning. 
- Sau đó, chúng ta xây dựng nhiệm vụ dự đoán loài Iris mà một loài hoa cụ thể thuộc về bằng cách sử dụng các phép đo vật lý của hoa. Chúng ta đã sử dụng một bộ dữ liệu các phép đo được chú thích bởi một chuyên gia với các loài chính xác để xây dựng mô hình của chúng ta, làm cho việc này trở thành một nhiệm vụ học tập có giám sát.
Thực sự Machine  Learning rất lý thú, nó có giúp chúng ta nhiều trong việc nặng nhọc về khối lượng tính toán, độ phức tạp. Nó đem lại tri thức cho bạn trong một khối dữ liệu lớn bạn có, với tri thức đó bạn có thể làm những việc quan trọng và có ý nghĩa. 

`Chào mừng bạn đến với Machine Learning.`